packages/@react-stately/tabs/src/useTabListState.ts

```diff
@@ -56,7 +56,7 @@ export function useTabListState<T extends object>(props: TabListStateOptions<T>)
       selectionManager.setFocusedKey(selectedKey);
     }
     lastSelectedKey.current = selectedKey;
-  });
+  }, [collection]);
```

Comment:
Don't we need this to also update when `selectedKey` changes, or if `selectionManager.isEmpty` or a bunch of other conditions?

Reply:
@snowystinger Hm, what did you try for the previous collection? This passes the test for me at least. Maybe im misunderstanding something.

```tsx
useEffect(() => {
  lastCollectionRef.current = collection;
  if (collection === lastCollectionRef.current) {
    return;
  }
  //more
});
```

Reply:
yeah, I'd like to get more at the root cause

I'd hoped to change the first `if` in the effect to include
```
lastCollection.current !== collection
```
and then leave off the dependency array, but it's not working for some reason. probably should have opened as a draft, but I forgot

Reply:
@snowystinger That's weird, that works for me 🤷

Reply:
? tests all pass?
have I messed up my own logic?
```
  let lastSelectedKey = useRef(currentSelectedKey);
  let lastCollection = useRef(collection);
  useEffect(() => {
    // Ensure a tab is always selected (in case no selected key was specified or if selected item was deleted from collection)
    let selectedKey = currentSelectedKey;
    if (lastCollection.current !== collection && (selectionManager.isEmpty || selectedKey == null || !collection.getItem(selectedKey))) {
...
    lastSelectedKey.current = selectedKey;
    lastCollection.current = collection;
  });
```

Reply:
yes, all tests pass exactly like this 😅 Also did a clean install and cleared the parcel cache just to make sure...

Reply:
I get 8 failures... even on a brand new pull. lets see if CI disagrees with me.

Reply:
I see, running the whole test pipeline i also get 8 failures in `Tabs.test.js`. I was running this test file in isolation so it appears to be some sort of cascading failure.

Reply:
I don't think it's a cascade failure, we have more behavior tests in React Spectrum, those are the ones failing, for the reasons outlined by @devongovett 

The crux of the issue is that the selectedkey is set at the same time as the item with that key is added.
When the selectedkey changes (regardless of the collection changing), we need to check to see if we should set a forced default since "no tab selected" is not an option.
At this point though, the collection isn't updated, that takes one more render pass.

As a result, the new key doesn't exist when we go to check if we need to set a default, so we set a default which triggers onSelectionChange.
*Then* the collection is ready, but it's too late.

Reply:
I see now, thanks 👍